src/handlers/miner_hbbft_handler.erl

```diff
@@ -457,77 +482,190 @@ restore(OldState, NewState) ->
-                         f = F, signatures = Signatures, signatures_required = Threshold0}) ->
-    Threshold = case Phase of
-                    sig -> F + 1;
-                    gossip -> Threshold0
-                end,
+% @doc Converts a list of `{Address, Signature}' tuples into a a list of the
+% Address' position in the `Addresses' list.
+-spec sigs_to_positions(sigs(), [addr()]) -> [pos_integer()].
+sigs_to_positions(Addr2Sig, Addresses) ->
+    Addr2Pos = lists:zip(Addresses, lists:seq(1, length(Addresses))),
+    Positions = lists:map(fun({Addr, _S, _V}) ->
+                            %% find member index
+                            {_, Pos} = lists:keyfind(Addr, 1, Addr2Pos),
+                            Pos
+                    end,
+                    Addr2Sig),
+    lists:usort(Positions).
 
-    %% filter out any signatures that are invalid or are not for a member of this DKG and dedup
-    case blockchain_block:verify_signatures(blockchain_block:deserialize(Artifact),
-                                            Members,
-                                            Signatures,
-                                            Threshold) of
-        {true, ValidSignatures} ->
-            %% So, this is a little dicey, if we don't need all N signatures, we might have competing subsets
-            %% depending on message order. Given that the underlying artifact they're signing is the same though,
-            %% it should be ok as long as we disregard the signatures for testing equality but check them for validity
+-spec md_version(blockchain_ledger_v1:ledger()) -> map | tuple.
+md_version(Ledger) ->
+    case blockchain:config(?election_version, Ledger) of
+        {ok, N} when N >= 3 ->
+            map;
+        _ ->
+            tuple
+    end.
+
+-spec state_reset(hbbft:hbbft_data(), #state{}) -> #state{}.
+state_reset(HBBFT, #state{}=S) ->
+    S#state{
+        hbbft          = HBBFT,
+        sigs_valid     = [],
+        sigs_invalid   = [],
+        sigs_unchecked = [],
+        artifact       = undefined,
+        sig_phase      = unsent,
+        bba            = <<>>,
+        seen           = #{}
+    }.
+
+-spec state_sigs_input(sigs(), #state{}) ->
+    {
+        {done, sigs()} | {gossip, sigs()} | pending,
+        #state{}
+    }.
+state_sigs_input(SigsIn, #state{hbbft=HBBFT}=S0) ->
+    CurRound = hbbft:round(HBBFT),
+    S1 = state_sigs_add(SigsIn, S0),
+    S2 = state_sigs_retry_unchecked(S1),
+    case state_sigs_check_if_enough(S2) of
+        {{some, {gossip, _}=Gossip}, S3} ->
+            {Gossip, S3#state{sig_phase = gossip}};
+        {
+            {some, {done, Sigs}=Done},
+            #state{artifact=Art, last_round_signed=LastRoundSigned}=S3
+        } when CurRound > LastRoundSigned ->
+            %% no point in doing this more than once
+            %% XXX What if artifact is undefined?
+            ok = miner:signed_block(Sigs, Art),
```

Comment:
@Vagabond If this `Art` is `undefined` - `miner:signed_block` will crash on `blockchain_block:deserialize(BinBlock)`, if I'm reading the call chain correctly.